imap_processing/ultra/l1b/de.py

```diff
@@ -23,35 +38,113 @@ def calculate_de(de_dataset, name):
-    de_dict["tof_start_stop"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["tof_stop_coin"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["tof_corrected"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["eventtype"] = np.zeros(len(epoch), dtype=np.uint64)
-    de_dict["vx_ultra"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vy_ultra"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vz_ultra"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["energy"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["species"] = np.zeros(len(epoch), dtype=np.uint64)
-    de_dict["event_efficiency"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vx_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vy_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vz_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vx_dps_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vy_dps_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vz_dps_sc"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vx_dps_helio"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vy_dps_helio"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["vz_dps_helio"] = np.zeros(len(epoch), dtype=np.float64)
-    de_dict["eventtimes"] = np.zeros(len(epoch), dtype=np.float64)
+    xf = get_front_x_position(
+        de_dataset["START_TYPE"].data,
+        de_dataset["START_POS_TDC"].data,
+    )
+
+    # Pulse height
+    ph_indices = np.where(
+        (de_dataset["STOP_TYPE"] == 1) | (de_dataset["STOP_TYPE"] == 2)
+    )[0]
+    ph_tof, ph_t2, ph_xb, ph_yb = get_ph_tof_and_back_positions(
+        de_dataset, xf, "ultra45"
+    )
+    ph_d, ph_yf = get_front_y_position(de_dataset["START_TYPE"].data[ph_indices], ph_yb)
+    ph_energy = get_energy_pulse_height(
+        de_dataset["STOP_TYPE"].data[ph_indices], ph_xb, ph_yb
+    )
+    ph_r = get_path_length((xf[ph_indices], ph_yf), (ph_xb, ph_yb), ph_d)
+    ph_ctof, ph_bin = determine_species_pulse_height(ph_energy, ph_tof, ph_r)
+    ph_etof, ph_xc = get_coincidence_positions(de_dataset, ph_tof, "ultra45")
+
+    # SSD
+    ssd_indices = np.where(de_dataset["STOP_TYPE"] >= 8)[0]
+    ssd_yb, _, _ = get_ssd_offset_and_positions(de_dataset)
+    ssd_d, ssd_yf = get_front_y_position(
+        de_dataset["START_TYPE"].data[ssd_indices], ssd_yb
+    )
+    ssd_tof, ssd = get_ssd_tof(de_dataset, xf)
+    ssd_energy = get_energy_ssd(de_dataset, ssd)
+    ssd_xb = np.zeros(len(ssd_yb))
+    ssd_r = get_path_length((xf[ssd_indices], ssd_yf), (ssd_xb, ssd_yb), ssd_d)
+    ssd_ctof, ssd_bin = determine_species_ssd(ssd_energy, ssd_tof, ssd_r)
+
+    # Combine ph_yb and ssd_yb along with their indices
+    combined_indices = np.concatenate((ph_indices, ssd_indices))
+
+    de_dict["x_front"] = xf
+
+    yb = np.concatenate((ph_yb, ssd_yb))
+    de_dict["y_back"] = yb[np.argsort(combined_indices)]
```

Comment:
It looks like you could store the results from `np.argsort(combined_indices)` and use that instead of having numpy perform the sorting algorithm so many times.